README.md
[![developed_using](https://img.shields.io/badge/developed%20using-Jetbrains%20Goland-lightgrey)](https://www.jetbrains.com/go/)
<br/>
![GitHub](https://img.shields.io/github/license/petrjahoda/display_webservice)
[![GitHub last commit](https://img.shields.io/github/last-commit/petrjahoda/display_webservice)](https://github.com/petrjahoda/display_webservice/commits/master)
[![GitHub issues](https://img.shields.io/github/issues/petrjahoda/display_webservice)](https://github.com/petrjahoda/display_webservice/issues)
<br/>
![GitHub language count](https://img.shields.io/github/languages/count/petrjahoda/display_webservice)
![GitHub top language](https://img.shields.io/github/languages/top/petrjahoda/display_webservice)
![GitHub repo size](https://img.shields.io/github/repo-size/petrjahoda/display_webservice)
<br/>
[![Docker Pulls](https://img.shields.io/docker/pulls/petrjahoda/display_webservice)](https://hub.docker.com/r/petrjahoda/display_webservice)
[![Docker Image Size (latest by date)](https://img.shields.io/docker/image-size/petrjahoda/display_webservice?sort=date)](https://hub.docker.com/r/petrjahoda/display_webservice/tags)
<br/>
[![developed_using](https://img.shields.io/badge/database-PostgreSQL-red)](https://www.postgresql.org) [![developed_using](https://img.shields.io/badge/runtime-Docker-red)](https://www.docker.com)

# Display WebService

## Description
Go web service, that shows web pages on port 81
* `/display_1` shows all workplaces with their statuses in small tiles

## Installation Information
Install under docker runtime using [this dockerfile image](https://github.com/petrjahoda/system/tree/master/latest) with this command: ```docker-compose up -d```

## Implementation Information
Check the software running with this command: ```docker stats```. <br/>
Display_webservice has to be running.

## Developer Information
Use software only as a [part of a system](https://github.com/petrjahoda/system) using Docker runtime.<br/>
 Do not run under linux, windows or mac on its own.

© 2020 Petr Jahoda

Example
![Display](screenshots/Screenshot%202021-08-26%20at%2010.49.28.png)


CHANGELOG.md
# Display WebService Changelog

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).

Please note, that this project, while following numbering syntax, it DOES NOT adhere
to [Semantic Versioning](http://semver.org/spec/v2.0.0.html) rules.

## Types of changes

* ```Added``` for new features.
* ```Changed``` for changes in existing functionality.
* ```Deprecated``` for soon-to-be removed features.
* ```Removed``` for now removed features.
* ```Fixed``` for any bug fixes.
* ```Security``` in case of vulnerabilities.

## [2022.2.2.18] - 2022-05-18

### Changed

- updated checkDatabaseConnection()

## [2022.2.1.28] - 2022-04-28

### Changed

- updated for creating docker images on apple silicon

## [2021.4.2.1] - 2021-11-01

### Changed

- latest modules

## [2021.4.1.28] - 2021-10-28

### Changed

- html rendering
- checking license

## [2021.4.1.4] - 2021-10-04

### Added

- updated to Go 1.17.1
- updated to latest modules

## [2021.3.3.15] - 2021-09-15

### Added

- license validation

## [2021.3.2.25] - 2021-08-25

### Changed

- font to Milliard

### Added

- css settings for @page

## [2021.3.2.19] - 2021-08-19

### Changed

- updated to latest go 1.17
- updated to latest go libraries

## [2021.2.3.14] - 2021-06-14

### Added

- copyright
- updated libraries

## [2021.2.2.13] - 2021-05-13

### Changed

- updated to latest go 1.16.4
- updated to latest go libraries

## [2021.2.2.3] - 2021-05-03

### Changed

- updated to latest go
- updated to latest go libraries

## [2021.2.1.8] - 2021-04-08

### Added

- application name to sql connection string

## [2021.2.1.7] - 2021-04-07

### Changed

- simplified streaming
- proper color from database

## [2021.2.1.1] - 2021-04-01

### Fixed

- closing sql connection

## [2021.1.3.30] - 2021-03-30

### Changed

- updated to latest go
- updated to latest libraries

## [2021.1.3.18] - 2021-03-18

### Changed

- updated to latest go
- updated to latest libraries

## [2021.1.2.21] - 2021-02-21

### Changed

- updated to latest go
- updated to latest libraries

## [2020.4.3.14] - 2020-12-14

### Changed

- updated to latest go
- updated to latest libraries

## [2020.4.2.20] - 2020-11-20

### Changed

- CSS tiles formatting

## [2020.4.2.17] - 2020-11-17

### Changed

- CSS formatting
- loading data on start

## [2020.4.2.17] - 2020-11-17

### Changed

- updated all go libraries
- updated Dockerfile
- updated create.sh
- metroui change to specific js and css
- font changed to proximanova

### Removed

- metroui css and js
- display_2

## [2020.4.1.26] - 2020-10-26

### Fixed

- fixed leaking goroutine bug when opening sql connections, the right way is this way

## [2020.3.3.28] - 2020-09-28

### Changed

- updated readme.md
- proper data loaded when page loaded

## [2020.3.3.27] - 2020-09-27

### Added

- screenshot to readme.md

## [2020.3.3.26] - 2020-09-26

### Changed

- updated to latest libraries

### Fixed

- proper loading metro.min.js

## [2020.3.2.22] - 2020-08-29

### Changed

- functions naming changed to idiomatic go (ThisFunction -> thisFunction)

## [2020.3.2.22] - 2020-08-22

### Added

- automatic go get -u all when creating docker image

## [2020.3.2.13] - 2020-08-13

### Changed

- updated to latest libraries
- updated to go 1.15 -> final program size reduced to 90%

## [2020.3.2.9] - 2020-08-09

### Added

- checking fro timezone until something is returned from database

## [2020.3.2.5] - 2020-08-05

### Added

- MIT license

### Changed

- code moved to more files
- logging updated
- name of functions updated

## [2020.3.2.4] - 2020-08-04

### Changed

- updated html files

### Fixed

- proper displaying order, downtime and user

## [2020.3.1.30] - 2020-07-30

### Changed

- update to latest database changes

## [2020.3.1.30] - 2020-07-30

### Fixed

- proper closing database connections with sqlDB, err := db.DB() and defer sqlDB.Close()

### Changed

- update to latest database changes
- dockerfile changed to "from scratch"

## [2020.3.1.29] - 2020-07-29

### Fixed

- proper checking for location while streaming time

## [2020.3.1.28] - 2020-07-28

### Added

- loading location from database for displaying

### Changed

- changed to gorm v2
- postgres only

### Removed

- logging to file
- all about config file

## [2020.2.2.18] - 2020-05-18

### Added

- init for actual service directory
- db.logmode(false)

## [2020.1.3.31] - 2020-03-31

### Added

- updated create.sh for uploading proper docker version automatically

### Fixed

- proper sending time to display

## [2020.1.3.5] - 2020-03-05

### Added

- webservice running on port 81
- /display_1 shows all workplaces with their status in small tiles
- /display_2 shows all workplaces with their status in wide tiles

main.go
package main

import (
	"encoding/json"
	"gorm.io/gorm/logger"
	"net/http"
	"os"
	"strconv"
	"time"

	"github.com/hyperboloide/lk"
	"github.com/julienschmidt/httprouter"
	"github.com/julienschmidt/sse"
	"github.com/kardianos/service"
	"github.com/petrjahoda/database"
	"gorm.io/driver/postgres"
	"gorm.io/gorm"
)

const version = "2022.2.2.18"
const serviceName = "Display WebService"
const serviceDescription = "Display webpages, for use with big televisions and displays"
const config = "user=postgres password=pj79.. dbname=system host=localhost port=5432 sslmode=disable application_name=display_webservice"

type program struct{}

func main() {
	logInfo("MAIN", serviceName+" ["+version+"] starting...")
	logInfo("MAIN", "© "+strconv.Itoa(time.Now().Year())+" Petr Jahoda")
	serviceConfig := &service.Config{
		Name:        serviceName,
		DisplayName: serviceName,
		Description: serviceDescription,
	}
	prg := &program{}
	s, err := service.New(prg, serviceConfig)
	if err != nil {
		logError("MAIN", "Cannot start: "+err.Error())
	}
	err = s.Run()
	if err != nil {
		logError("MAIN", "Cannot start: "+err.Error())
	}
}

func (p *program) Start(service.Service) error {
	logInfo("MAIN", serviceName+" ["+version+"] started")
	go p.run()
	return nil
}

func (p *program) Stop(service.Service) error {
	logInfo("MAIN", serviceName+" ["+version+"] stopped")
	return nil
}

func (p *program) run() {
	checkDatabaseConnection()
	programIsActive := false
	updateProgramVersion()
	for !programIsActive {
		programIsActive = checkActivation(programIsActive)
		if !programIsActive {
			logInfo("MAIN", serviceName+": licence is not valid")
			time.Sleep(10 * time.Second)
		}
	}
	router := httprouter.New()
	timer := sse.New()
	workplaces := sse.New()
	overview := sse.New()
	router.ServeFiles("/js/*filepath", http.Dir("js"))
	router.ServeFiles("/css/*filepath", http.Dir("css"))
	router.ServeFiles("/fonts/*filepath", http.Dir("fonts"))

	router.GET("/display_1", display1)

	router.Handler("GET", "/time", timer)
	router.Handler("GET", "/workplaces", workplaces)
	router.Handler("GET", "/overview", overview)
	timezone := readTimeZoneFromDatabase()
	go streamTime(timer, timezone)
	go streamWorkplaces(workplaces)
	go streamOverview(overview)
	err := http.ListenAndServe(":81", router)
	if err != nil {
		logError("MAIN", "Problem starting service: "+err.Error())
		os.Exit(-1)
	}
	logInfo("MAIN", serviceName+" ["+version+"] running")
}

func checkActivation(programIsActive bool) bool {
	if programIsActive {
		return true
	}
	db, err := gorm.Open(postgres.Open(config), &gorm.Config{})
	sqlDB, _ := db.DB()
	defer sqlDB.Close()
	if err != nil {
		return false
	}
	var customerName database.Setting
	var softwareLicense database.Setting
	db.Where("name = ?", "company").Find(&customerName)
	db.Where("name = ?", serviceName).Find(&softwareLicense)
	const publicKeyBase32 = "ARIVIK3FHZ72ERWX6FQ6Z3SIGHPSMCDBRCONFKQRWSDIUMEEESQULEKQ7J7MZVFZMJDFO6B46237GOZETQ4M2NE32C3UUNOV5EUVE3OIV72F5LQRZ6DFMM6UJPELARG7RLJWKQRATUWD5YT46Q2TKQMPPGIA===="
	publicKey, err := lk.PublicKeyFromB32String(publicKeyBase32)
	if err != nil {
		return false
	}
	license, err := lk.LicenseFromB32String(softwareLicense.Note)
	if err != nil {
		return false
	}
	if ok, err := license.Verify(publicKey); err != nil {
		return false
	} else if !ok {
		return false
	}
	result := struct {
		Software string `json:"software"`
		Customer string `json:"customer"`
	}{}
	if err := json.Unmarshal(license.Data, &result); err != nil {
		return false
	}
	if result.Customer == customerName.Value && result.Software == softwareLicense.Name {
		logInfo("MAIN", serviceName+": licence is valid")
		return true
	}
	return false
}

func checkDatabaseConnection() {
	databaseConnected := false
	for !databaseConnected {
		db, err := gorm.Open(postgres.Open(config), &gorm.Config{Logger: logger.Default.LogMode(logger.Silent)})
		sqlDB, _ := db.DB()
		if err != nil {
			logError("SYSTEM", "Database not connected: "+err.Error())
			time.Sleep(1 * time.Second)
		} else {
			var checkUser database.User
			db.Where("email = ?", "admin@admin.com").Find(&checkUser)
			if checkUser.ID == 0 {
				logError("SYSTEM", "Database not initialized")
				sqlDB.Close()
				time.Sleep(1 * time.Second)
			} else {
				sqlDB.Close()
				databaseConnected = true
			}
		}
	}
}


display.go
package main

import (
	"github.com/goodsign/monday"
	"github.com/julienschmidt/httprouter"
	"github.com/petrjahoda/database"
	"gorm.io/driver/postgres"
	"gorm.io/gorm"
	"html/template"
	"net/http"
	"strconv"
	"time"
)

type LcdWorkplaces struct {
	ProductionPercent string
	DowntimePercent   string
	OfflinePercent    string
	ProductionColor   string
	DowntimeColor     string
	PowerOffColor     string
	Time              string
	Version           string
	LcdWorkplaces     []LcdWorkplace
}

type LcdWorkplace struct {
	StateColor      string
	StateColorStyle string
	Name            string
	User            string
	Order           string
	Downtime        string
	Duration        string
}

func display1(writer http.ResponseWriter, _ *http.Request, _ httprouter.Params) {
	logInfo("HTML", "Display 1 process started")
	timer := time.Now()
	tmpl := template.Must(template.ParseFiles("html/display_1.gohtml"))
	var workplaces []database.Workplace
	lcdWorkplaces := LcdWorkplaces{}
	db, err := gorm.Open(postgres.Open(config), &gorm.Config{})
	sqlDB, _ := db.DB()
	defer sqlDB.Close()
	if err != nil {
		logError("HTML", "Problem opening database: "+err.Error())
	}
	db.Order("Name asc").Find(&workplaces)
	location := downloadActualLocation()
	loc, _ := time.LoadLocation(location)
	for _, workplace := range workplaces {
		stateRecord := database.StateRecord{}
		db.Where("workplace_id = ?", workplace.ID).Last(&stateRecord)
		state := database.State{}
		db.Where("id = ?", stateRecord.StateID).Find(&state)
		orderRecord := database.OrderRecord{}
		db.Where("workplace_id = ?", workplace.ID).Where("date_time_end is null").Last(&orderRecord)
		workplaceHasOpenOrder := orderRecord.ID > 0
		downtimeRecord := database.DowntimeRecord{}
		db.Where("workplace_id = ?", workplace.ID).Where("date_time_end is null").Last(&downtimeRecord)
		downtime := database.Downtime{}
		db.Where("id = ?", downtimeRecord.DowntimeID).Find(&downtime)
		var userName string
		order := database.Order{}
		if workplaceHasOpenOrder {
			db.Where("id = ?", orderRecord.OrderID).Find(&order)
			user := database.User{}
			db.Where("id = ?", orderRecord.UserId.Int32).Find(&user)
			userName = user.FirstName + " " + user.SecondName
		}
		color := "green"
		colorStyle := "background: " + state.Color

		duration := time.Now().In(loc).Sub(stateRecord.DateTimeStart.In(loc)).Round(1 * time.Minute).String()
		if err != nil {
			logError(workplace.Name, "Problem parsing datetime: "+err.Error())
		}
		logInfo(workplace.Name, duration)
		logInfo(workplace.Name, "Workplace color: "+color+", order: "+order.Name+", downtime: "+downtime.Name+", user: "+userName)
		logInfo("HTML", "Adding workplace: "+workplace.Name)
		lcdWorkplace := LcdWorkplace{Name: workplace.Name, User: userName, StateColor: color, Duration: duration, Downtime: downtime.Name, Order: order.Name, StateColorStyle: colorStyle}
		lcdWorkplaces.LcdWorkplaces = append(lcdWorkplaces.LcdWorkplaces, lcdWorkplace)
	}
	var states []database.State
	db.Find(&states)
	db.Find(&workplaces)
	production := 0
	downtime := 0
	offline := 0
	for _, workplace := range workplaces {
		stateRecord := database.StateRecord{}
		db.Where("workplace_id = ?", workplace.ID).Last(&stateRecord)
		switch stateRecord.StateID {
		case 1:
			production++
		case 2:
			downtime++
		case 3:
			offline++
		}
	}
	sum := production + offline + downtime
	productionPercent := production * 100 / sum
	downtimePercent := downtime * 100 / sum
	offlinePercent := 100 - productionPercent - downtimePercent
	floatPointMiscalculation := offline == 0 && offlinePercent > 0
	if floatPointMiscalculation {
		offlinePercent = 0
		if downtimePercent > productionPercent {
			downtimePercent++
		} else {
			productionPercent++
		}
	}
	lcdWorkplaces.ProductionPercent = "Production " + strconv.Itoa(productionPercent) + "%"
	lcdWorkplaces.DowntimePercent = "Downtime " + strconv.Itoa(downtimePercent) + "%"
	lcdWorkplaces.OfflinePercent = "Poweroff " + strconv.Itoa(offlinePercent) + "%"
	for _, state := range states {
		switch state.Name {
		case "Production":
			lcdWorkplaces.ProductionColor = state.Color
		case "Downtime":
			lcdWorkplaces.DowntimeColor = state.Color
		case "Poweroff":
			lcdWorkplaces.PowerOffColor = state.Color
		}
	}
	lcdWorkplaces.Version = "version: " + version
	timezone := readTimeZoneFromDatabase()
	if timezone == "" {
		timezone = readTimeZoneFromDatabase()
	}
	if err != nil {
		logError("MAIN", "Problem loading location: "+timezone)
	} else {
		lcdWorkplaces.Time = monday.Format(time.Now().In(loc), "Monday, 2. January 2006, 15:04:05", monday.LocaleEnUS)
	}
	_ = tmpl.Execute(writer, lcdWorkplaces)
	logInfo("HTML", "Display 1 process ended in "+time.Since(timer).String())
}


js/system.js
const time = new EventSource('/time');
time.addEventListener('time', (e) => {
    document.getElementById("time").innerHTML = e.data;

}, false);
const overview = new EventSource('/overview');
overview.addEventListener('overview', (e) => {
    const overviewData = e.data.split(";");
    document.getElementById("production").innerHTML = overviewData[0];
    document.getElementById("downtime").innerHTML = overviewData[1];
    document.getElementById("offline").innerHTML = overviewData[2];

}, false);
const workplaces = new EventSource('/workplaces');
workplaces.addEventListener('workplaces', (e) => {
    const overviewData = e.data.split(";");
    document.getElementById(overviewData[0]).innerHTML = overviewData[1];
    document.getElementById(overviewData[0]).style.background = overviewData[2]
}, false);

service.go
package main

import (
	"github.com/goodsign/monday"
	"github.com/julienschmidt/sse"
	"github.com/petrjahoda/database"
	"gorm.io/driver/postgres"
	"gorm.io/gorm"
	"strconv"
	"time"
)

func readTimeZoneFromDatabase() string {
	logInfo("MAIN", "Reading timezone from database")
	timer := time.Now()
	db, err := gorm.Open(postgres.Open(config), &gorm.Config{})
	sqlDB, _ := db.DB()
	defer sqlDB.Close()
	if err != nil {
		logError("MAIN", "Problem opening database: "+err.Error())
		return ""
	}
	var settings database.Setting
	db.Where("name=?", "timezone").Find(&settings)
	logInfo("MAIN", "Timezone read in "+time.Since(timer).String())
	return settings.Value
}

func updateProgramVersion() {
	logInfo("MAIN", "Writing program version into settings")
	timer := time.Now()
	db, err := gorm.Open(postgres.Open(config), &gorm.Config{})
	sqlDB, _ := db.DB()
	defer sqlDB.Close()
	if err != nil {
		logError("MAIN", "Problem opening database: "+err.Error())
		return
	}
	var existingSettings database.Setting
	db.Where("name=?", serviceName).Find(&existingSettings)
	existingSettings.Name = serviceName
	existingSettings.Value = version
	db.Save(&existingSettings)
	logInfo("MAIN", "Program version written into settings in "+time.Since(timer).String())
}

func streamOverview(streamer *sse.Streamer) {
	for {
		logInfo("SSE", "Streaming overview started")
		timer := time.Now()
		production, downtime, offline := downloadDataForStreaming()
		sum := production + offline + downtime
		if sum == 0 {
			streamer.SendString("", "overview", "Production 0%;Downtime 0%;Poweroff 0%")
			continue
		} else {
			productionPercent := production * 100 / sum
			downtimePercent := downtime * 100 / sum
			offlinePercent := 100 - productionPercent - downtimePercent
			floatPointMiscalculation := offline == 0 && offlinePercent > 0
			if floatPointMiscalculation {
				offlinePercent = 0
				if downtimePercent > productionPercent {
					downtimePercent++
				} else {
					productionPercent++
				}
			}
			streamer.SendString("", "overview", "Production "+strconv.Itoa(productionPercent)+"%;Downtime "+strconv.Itoa(downtimePercent)+"%;Poweroff "+strconv.Itoa(offlinePercent)+"%")
		}
		logInfo("SSE", "Streaming overview ended in "+time.Since(timer).String())
		time.Sleep(10 * time.Second)
	}
}

func downloadDataForStreaming() (int, int, int) {
	db, err := gorm.Open(postgres.Open(config), &gorm.Config{})
	sqlDB, _ := db.DB()
	defer sqlDB.Close()
	if err != nil {
		logError("SSE", "Problem opening database: "+err.Error())
		return 0, 0, 0
	}
	production := 0
	downtime := 0
	offline := 0
	var stateRecords []database.StateRecord
	// select * from (select state_records.id as id, states.name as state, workplaces.name as workplace from state_records join workplaces on state_records.workplace_id=workplaces.id join states on state_records.state_id=states.id) as results where id in (select distinct max(id) as id from state_records group by workplace_id)
	db.Raw("select * from state_records where id in (select distinct max(id) as id from state_records group by workplace_id)").Find(&stateRecords)
	for _, record := range stateRecords {
		switch record.StateID {
		case 1:
			production++
		case 2:
			downtime++
		case 3:
			offline++
		}
	}
	return production, downtime, offline
}

func streamWorkplaces(streamer *sse.Streamer) {
	for {
		logInfo("SSE", "Streaming workplaces started")
		timer := time.Now()
		location := downloadActualLocation()
		loc, _ := time.LoadLocation(location)
		workplaces := downloadActualWorkplaces()
		if len(workplaces) > 0 {
			cachedDowntimeRecords := downloadDowntimeRecords()
			cachedOrderRecords := downloadOrderRecords()
			cachedStateRecords := downloadStateRecords()
			cachedDowntimes := downloadDowntimes()
			cachedOrders := downloadOrders()
			cachedUsers := downloadUsers()
			cachedStates := downloadStates()
			for _, workplace := range workplaces {
				workplaceDowntimeRecord := cachedDowntimeRecords[int(workplace.ID)]
				workplaceOrderRecord := cachedOrderRecords[int(workplace.ID)]
				workplaceStateRecord := cachedStateRecords[int(workplace.ID)]
				streamer.SendString("", "workplaces", workplace.Name+";<b>"+workplace.Name+"</b><br>"+cachedUsers[int(workplaceOrderRecord.UserId.Int32)].FirstName+" "+cachedUsers[int(workplaceOrderRecord.UserId.Int32)].SecondName+"<br>"+cachedOrders[workplaceOrderRecord.OrderID].Name+"<br>"+cachedDowntimes[workplaceDowntimeRecord.DowntimeID].Name+"<br><sub>"+time.Now().In(loc).Sub(workplaceStateRecord.DateTimeStart.In(loc)).Round(1*time.Minute).String()+"</sub>;"+cachedStates[workplaceStateRecord.StateID].Color)
			}
		}
		logInfo("SSE", "Streaming workplaces ended in "+time.Since(timer).String())
		time.Sleep(10 * time.Second)
	}
}

func downloadStateRecords() map[int]database.StateRecord {
	db, err := gorm.Open(postgres.Open(config), &gorm.Config{})
	sqlDB, _ := db.DB()
	defer sqlDB.Close()
	if err != nil {
		logError("SSE", "Problem opening database: "+err.Error())
		return nil
	}
	var stateRecords []database.StateRecord
	db.Raw("select * from state_records where id in (select distinct max(id) as id from state_records group by workplace_id)").Find(&stateRecords)
	cachedStateRecords := make(map[int]database.StateRecord)
	for _, stateRecord := range stateRecords {
		cachedStateRecords[stateRecord.WorkplaceID] = stateRecord
	}
	return cachedStateRecords
}

func downloadOrderRecords() map[int]database.OrderRecord {
	db, err := gorm.Open(postgres.Open(config), &gorm.Config{})
	sqlDB, _ := db.DB()
	defer sqlDB.Close()
	if err != nil {
		logError("SSE", "Problem opening database: "+err.Error())
		return nil
	}
	var orderRecords []database.OrderRecord
	db.Where("date_time_end is null").Find(&orderRecords)
	cachedOrderRecords := make(map[int]database.OrderRecord)
	for _, orderRecord := range orderRecords {
		cachedOrderRecords[orderRecord.WorkplaceID] = orderRecord
	}
	return cachedOrderRecords
}

func downloadStates() map[int]database.State {
	db, err := gorm.Open(postgres.Open(config), &gorm.Config{})
	sqlDB, _ := db.DB()
	defer sqlDB.Close()
	if err != nil {
		logError("SSE", "Problem opening database: "+err.Error())
		return nil
	}
	var states []database.State
	db.Find(&states)
	cachedStates := make(map[int]database.State)
	for _, state := range states {
		cachedStates[int(state.ID)] = state
	}
	return cachedStates
}

func downloadUsers() map[int]database.User {
	db, err := gorm.Open(postgres.Open(config), &gorm.Config{})
	sqlDB, _ := db.DB()
	defer sqlDB.Close()
	if err != nil {
		logError("SSE", "Problem opening database: "+err.Error())
		return nil
	}
	var users []database.User
	db.Find(&users)
	cachedUsers := make(map[int]database.User)
	for _, user := range users {
		cachedUsers[int(user.ID)] = user
	}
	return cachedUsers
}

func downloadOrders() map[int]database.Order {
	db, err := gorm.Open(postgres.Open(config), &gorm.Config{})
	sqlDB, _ := db.DB()
	defer sqlDB.Close()
	if err != nil {
		logError("SSE", "Problem opening database: "+err.Error())
		return nil
	}
	var orders []database.Order
	db.Find(&orders)
	cachedOrders := make(map[int]database.Order)
	for _, order := range orders {
		cachedOrders[int(order.ID)] = order
	}
	return cachedOrders
}

func downloadDowntimes() map[int]database.Downtime {
	db, err := gorm.Open(postgres.Open(config), &gorm.Config{})
	sqlDB, _ := db.DB()
	defer sqlDB.Close()
	if err != nil {
		logError("SSE", "Problem opening database: "+err.Error())
		return nil
	}
	var downtimes []database.Downtime
	db.Find(&downtimes)
	cachedDowntimes := make(map[int]database.Downtime)
	for _, downtime := range downtimes {
		cachedDowntimes[int(downtime.ID)] = downtime
	}
	return cachedDowntimes
}

func downloadDowntimeRecords() map[int]database.DowntimeRecord {
	db, err := gorm.Open(postgres.Open(config), &gorm.Config{})
	sqlDB, _ := db.DB()
	defer sqlDB.Close()
	if err != nil {
		logError("SSE", "Problem opening database: "+err.Error())
		return nil
	}
	var downtimeRecords []database.DowntimeRecord
	db.Where("date_time_end is null").Find(&downtimeRecords)
	cachedDowntimeRecords := make(map[int]database.DowntimeRecord)
	for _, downtimeRecord := range downtimeRecords {
		cachedDowntimeRecords[downtimeRecord.WorkplaceID] = downtimeRecord
	}
	return cachedDowntimeRecords
}

func downloadActualLocation() string {
	db, err := gorm.Open(postgres.Open(config), &gorm.Config{})
	sqlDB, _ := db.DB()
	defer sqlDB.Close()
	if err != nil {
		logError("SSE", "Problem opening database: "+err.Error())
		return ""
	}
	var timezone database.Setting
	db.Where("name=?", "timezone").Find(&timezone)
	return timezone.Value
}

func downloadActualWorkplaces() []database.Workplace {
	db, err := gorm.Open(postgres.Open(config), &gorm.Config{})
	sqlDB, _ := db.DB()
	defer sqlDB.Close()
	if err != nil {
		logError("SSE", "Problem opening database: "+err.Error())
		return nil
	}
	var workplaces []database.Workplace
	db.Find(&workplaces)
	return workplaces
}

func streamTime(streamer *sse.Streamer, timezone string) {
	logInfo("SSE", "Streaming time process started")
	for {
		if timezone == "" {
			timezone = readTimeZoneFromDatabase()
		}
		location, err := time.LoadLocation(timezone)
		if err != nil {
			logError("MAIN", "Problem loading location: "+timezone)
		} else {
			streamer.SendString("", "time", monday.Format(time.Now().In(location), "Monday, 2. January 2006, 15:04:05", monday.LocaleEnUS))
			time.Sleep(1 * time.Second)
		}
	}
}
